infrastructure/proto/log_replication_transport.proto

```diff
@@ -7,16 +7,20 @@ import "google/protobuf/any.proto";
 import "service/corfu_message.proto";
 
 
-service LogReplicationChannel {
+service LogReplication {
 
-    // Client sends Negotiation Request
+    // Source sends Negotiation Request
     rpc negotiate(org.corfudb.runtime.RequestMsg) returns (org.corfudb.runtime.ResponseMsg) {}
 
-    // Client sends Negotiation Request
+    // Connection initiator sends Leadership Request
     rpc queryLeadership(org.corfudb.runtime.RequestMsg) returns (org.corfudb.runtime.ResponseMsg) {}
 
-    // A client-side streaming RPC where the client writes a sequence of messages and sends them to the server,
-    // again using a provided stream. Once the client has finished writing the messages, it waits for the server
-    // to read them all and return its response (ack).
+    // Used only by Source when connection starter: A client-side streaming RPC where the client writes a sequence of
+    // messages and sends them to the server, again using a provided stream. Once the client has finished writing the
+    // messages, it waits for the server to read them all and return its response (ack).
     rpc replicate(stream org.corfudb.runtime.RequestMsg) returns (org.corfudb.runtime.ResponseMsg) {}
+
+    // Used only by Sink when connection starter. This RPC creates a long living stream, on which the source and
+    // sink exchange negotiation-msg/replication-msg/ACKs
+    rpc subscribeAndStartReplication(stream org.corfudb.runtime.ResponseMsg) returns (stream org.corfudb.runtime.RequestMsg) {}
```

Comment:
Got it a bit, is it because `ClientChannelAdapter` is only created at connection initiator and `ServerChannelAdapter` is only created on receiver? but now we are now having router per session so a cluster could be initiator and receiver simultaneously right? 

I thought would it be possible/cleaner to let transport layer agnostic of Source/Sink by just  creating both `ClientChannelAdapter` and `ServerChannelAdapter` for every cluster? such that we can avoid this reversed usage of request/response and let upper level decide which RPC it wants to invoke

Reply:
1. The rpc is the other way round...It is sending a response and receiving a request.
2. the "lr_subscribe_request" on L209 of VerifyRemoteSinkLeader, is a proto msg name...I agree it can be renamed. how about "lr_subscribe_msg" ?
3. This rpc is needed because GRPC has a constraint that only the client can setup a channel and a stream and the server can send msgs via that stream only. 
In the case that SINK is the connection starter, it needs to set up a long living stream so that SOURCE continues to drive the replication. In this case, SOURCE cannot create its own stream and is dependent on the SIINK to create a stream.
"stream" here can be thought of as a pipe between client-server for an rpc (msg exchanges).  There is one stream for every rpc.

Reply:
I see the constraints now, so when Source is connection receiver, it use this channel to send all the RPC "requests" via `ServerChannelAdapter` in a form of "response" right?

Reply:
Discussed offline, clear now